### PR TITLE
license: support CAL license

### PR DESCRIPTION
**Pull request by hardening**
This patch simplifies the licensing code mutualizing encryption / decryption routines. It also adds the support for client_info packet that allows to send a previously saved CAL file.

### CODE REVIEW

**Comment by bmiklautz** [comment]
@hardening what servers have you tested against? I'll do some testing against our test setup as well.

**Comment by hardening** [comment]
I guess any server configured with license server configured in per device mode should trigger the added code.

**Comment by hardening** [comment]
Hum thinking of it perhaps in per user mode it is triggered too, not tested that.

**Comment by hardening** [comment]
@akallabeth we've discussed of that with @bmiklautz (privately). The changes are useful with a server license configured in per device mode (the server seems to link the MAC address and the CAL file).

**Comment by akallabeth** [comment]
@hardening then I'll check with @bmiklautz tomorrow and approve it ;)

**Comment by akallabeth** [comment]
@hardening nice improvement, btw

**Comment by hardening** [comment]
option for old behaviour added (off by default)

**Comment by bmiklautz** [comment]
@hardening thanks. Looks good for me. When we do have a second review I opt in to merge it.

File: libfreerdp/core/license.c [hunk 1]
```diff
@@ -132,21 +139,178 @@ static void license_print_product_info(LICENSE_PRODUCT_INFO* productInfo)
...
+
+	if (written != 1)
+	{
+		DeleteFile(filepathNew);
+		goto fail_file_open;
+	}
+
+	ret = MoveFileEx(filepathNew, filepath, MOVEFILE_REPLACE_EXISTING);
+
+fail_file_open:
```
Comment (akallabeth) [inline libfreerdp/core/license.c]: initializing variables with `NULL` could avoid having more than one `goto` label

File: libfreerdp/core/license.c [hunk 2]
```diff
@@ -132,21 +139,178 @@ static void license_print_product_info(LICENSE_PRODUCT_INFO* productInfo)
...
+		goto error_read;
+
+	*dataLen = length;
+
+	fclose(fp);
+	free(calPath);
+	free(licenseStorePath);
+	return ret;
+
+error_read:
```
Comment (akallabeth) [inline libfreerdp/core/license.c]: initializing variables with `NULL` can avoid having to use multiple `goto` labels

**Comment by hardening** [comment]
I think `LICENSE_NULL_CLIENT_RANDOM` and `LICENSE_NULL_PREMASTER_SECRET`are flags to behave like `rdesktop` (yes it's not secure on the license part).

File: libfreerdp/core/license.c (hunk 1 above)
Comment (hardening) [reply libfreerdp/core/license.c]: good catch, changing this.

File: libfreerdp/core/license.c (hunk 2 above)
Comment (hardening) [reply libfreerdp/core/license.c]: here I can simplify a little, but not completely getting rid of the 2 paths

**Comment by hardening** [comment]
@akallabeth I've fixed what you pointed out

**Pull request merged**